test/sqllogictest/transform/table_elimination.slt

```diff
@@ -0,0 +1,19 @@
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
```

Comment:
The multi-input union case that you suggested is not optimized, the union is not perfectly balanced:

```
materialize=> explain (select * from t1 union all select * from t1 union all select * from t1) except all (select * from t1 union all select * from t1 union all select * from t1);
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %1 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %2 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %3 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %4 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %5 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %6 =                            +
 | Union %3 %4 %5                +
 | Negate                        +
                                 +
 %7 =                            +
 | Union %0 %1 %2 %6             +
 | Threshold
(1 row)
```

Reply:
Hmm, I see this as a problem in that Unions are not properly fused. I think in a follow-on PR, Union fusion (which is in `src/transform/src/fusion/union.rs`) should be modified so that if you have a Negate around a Union, the Negate gets pushed down to each input. Then the Unions can be fused.
```
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %1 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %2 =                            +
 | Get materialize.public.t1 (u5)+
                                 +
 %3 =                            +
 | Get materialize.public.t1 (u5)+
 | Negate
                                 +
 %4 =                            +
 | Get materialize.public.t1 (u5)+
 | Negate
                                 +
 %5 =                            +
 | Get materialize.public.t1 (u5)+
 | Negate
                                 +
 %7 =                            +
 | Union %0 %1 %2 %3 %4 %5              +
 | Threshold

```

Reply:
I've implemented this change in a separate branch on top of these changes that I will post a PR for once the naming issue here gets resolved, so I can rebase that other branch before publishing the PR.

Reply:
Addressed by #6474